src/palace/manager/integration/license/opds/opds2/api.py

```diff
@@ -48,28 +74,155 @@ def __init__(self, _db: Session, collection: Collection):
+
+        entity_id = saml_subject.idp
+        person_scoped_affiliation = None
+
+        if (
+            saml_subject.attribute_statement
+            and saml_subject.attribute_statement.attributes
+        ):
+            scoped_affiliation_attr = saml_subject.attribute_statement.attributes.get(
+                SAMLAttributeType.eduPersonScopedAffiliation.name
+            )
+
+            if scoped_affiliation_attr:
+                person_scoped_affiliation = scoped_affiliation_attr.values
+
+        return entity_id, person_scoped_affiliation
+
+    def _build_template_parameters(
+        self, patron: Patron, datasource: DataSource, variable_names: MutableSet[str]
+    ) -> dict[str, str | list[str]]:
+        """Build template parameters based on what the template requires.
+
+        :param patron: The patron to get parameters for
+        :param datasource: The datasource for patron_id lookup
+        :param variable_names: Set of variable names from the URI template
+
+        :return: Dictionary of template parameters
+
+        :raises CannotFulfill: If template requires SAML parameters but patron lacks them
+        """
+        parameters: dict[str, str | list[str]] = {}
+
+        requested_variables = variable_names & SUPPORTED_TEMPLATE_VARIABLES
+
+        if TemplateVariable.PATRON_ID in requested_variables:
+            patron_id = patron.identifier_to_remote_service(datasource)
+            parameters[TemplateVariable.PATRON_ID] = patron_id
+
+        # Handle SAML template variables if any are requested.
+        # The logic here ensures that if the template requires SAML parameters, the patron
+        # must have ALL required SAML credentials. Partial data is not accepted - if the
+        # template asks for both entity_id and affiliation, the patron must have both.
+        if requested_saml_variables := (SAML_TEMPLATE_VARIABLES & requested_variables):
+            # Fetch patron's SAML credentials
+            entity_id, person_scoped_affiliation = (
+                self._get_saml_token_template_parameters(patron)
+            )
+
+            # Try to add each requested SAML variable to parameters
+            if TemplateVariable.SAML_ENTITY_ID in requested_variables and entity_id:
+                parameters[TemplateVariable.SAML_ENTITY_ID] = entity_id
+
+            if (
+                TemplateVariable.SAML_PERSON_SCOPED_AFFILIATION in requested_variables
+                and person_scoped_affiliation
+            ):
+                parameters[TemplateVariable.SAML_PERSON_SCOPED_AFFILIATION] = (
+                    person_scoped_affiliation
+                )
+
+            # Check if any requested SAML variables are still missing from parameters.
+            # This will be true if the patron lacks the required SAML credentials.
+            missing_saml_variables = requested_saml_variables - parameters.keys()
+            if missing_saml_variables:
```

Comment:
Minor: Could use ye olde Walrus operator here, but the linting might make that harder to read.

```suggestion
            if (
                missing_saml_variables := requested_saml_variables
                - parameters.keys()
            ):
```

Reply:
I initially had it using walrus, but it was really ugly after formatting, so I changed it back